### PR TITLE
Versioning 3

### PR DESCRIPTION
**Pull request by antlai-temporal**
## What was changed

This PR implements a new deployment based approach for Worker Versioning in go. 

This is an umbrella PR, with several PRs in the versioning-3 branch already reviewed:

#1675 
#1692 
#1708 
#1722
#1727
#1733
#1742 
 
It is marked as draft, missing some system tests, and waiting on a CLI release that provides the new server.

But even in draft mode, it should be reasonably stable, please give feedback... Thanks!

### CODE REVIEW

File: internal/deployment_client.go
Comment (Quinn-With-Two-Ns) [inline internal/deployment_client.go:166]:
```suggestion
		// of type *commonpb.Payload, the clients converter will be used to generate
```

File: internal/internal_workflow_execution_options.go
Comment (Quinn-With-Two-Ns) [inline internal/internal_workflow_execution_options.go:50]: Is there a reason we chose this approach to exclude fields over say making the members of `WorkflowExecutionOptions` pointers and treating `nil` as unset

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:166]: Thanks, done

File: internal/internal_workflow_execution_options.go
Comment (antlai-temporal) [reply internal/internal_workflow_execution_options.go:50]: I'm following closely the proto api that uses a FieldMask. In general, I've been trying to avoid pointers in the api, and be more explicit about what to change, it is sometimes easy to forget that nil means delete otherwise...

Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow_execution_options.go:50]: My suggestion was a `nil` pointer would leave a field unchanged , not clear it.

File: client/client.go
Comment (Quinn-With-Two-Ns) [inline client/client.go:956]: I fear this API may get confused with `UpdateWorkflow`, but nothing you can do in this PR as your just copying the naming from the grpc API

File: internal/schedule_client.go
Comment (Quinn-With-Two-Ns) [inline internal/schedule_client.go:287]: Nit:

```suggestion
		// To unset the override after the workflow is running, use [Client.UpdateWorkflowExecutionOptions].
```

There are a few other godocs that could be links

Comment (antlai-temporal) [reply internal/schedule_client.go:287]: Yes, I missed quite a few, I will made them links

File: internal/internal_workflow_execution_options.go
Comment (antlai-temporal) [reply internal/internal_workflow_execution_options.go:50]: How do you clear things? Do you use a tombstone?

Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow_execution_options.go:50]: Well here you say 

> // When it includes a field name, but the corresponding WorkflowExecutionOptions field has not been set,
		// it will remove previous overrides for that field.
		
So you are OK relying on the default value to mean clear the field , you can do the same if it takes a pointer.

Comment (antlai-temporal) [reply internal/internal_workflow_execution_options.go:50]: Sure, but I have the other safety net of listing keys that I will delete. I can see people reusing a data structure with the pointers, and making mistakes because they did not clear all of them... @cretz What do you think about a separate selector array vs pointers?

**Comment by Quinn-With-Two-Ns** [comment]
Overall the client API makes sense to me, I didn't review the underlying gRPC api to ensure the API is being invoked or mocked correctly.

File: internal/worker.go
Comment (ShahabT) [inline internal/worker.go:34]: Let's discuss these options through on Mon sync before merging to main.

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:40]: Can you confirm where this value is needed by users?

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:103]:
```suggestion
		Current bool
```

Pedantic though, don't need to change

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:140]: Is this needed?

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:155]: You are inconsistent in this type with whether you want to prefix the field name with `Deployment` when the type is prefixed with `Deployment`. But I can see a case for intentional inconsistency to disambiguate the multiple uses of the term "info".

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:106]:
```suggestion
		TaskQueueInfos []DeploymentTaskQueueInfo
```

I think this is clearer, or just `TaskQueues`

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:71]:
```suggestion
		// BuildID - Identifies the worker's code and configuration version.
```

Pedantic, but you are inconsistent with whether you capitalize the doc after ` - `

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:233]: Why would I ever call `Describe` when this has everything describe does and more? Why even have `Describe` or better yet, why not make `Describe` return reachability information?

File: internal/internal_deployment_client.go
Comment (cretz) [inline internal/internal_deployment_client.go:61]: Combine the conditionals into a single `if`

File: internal/internal_deployment_client.go
Comment (cretz) [inline internal/internal_deployment_client.go:63]: Can you confirm that the API will never return an empty list with a next page token present? Sometimes some list implementations with filters will apply the filter post-page fetch which means it could filter out all items on the page but still expect another page call.

File: internal/internal_deployment_client.go
Comment (cretz) [inline internal/internal_deployment_client.go:177]: Should add this to the top of `List` too

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:222]: How confident are you that we will never add anything else to `DescribeDeploymentResponse`?

File: internal/deployment_client.go
Comment (cretz) [inline internal/deployment_client.go:237]: How confident are you that we will never add anything else to `GetCurrentDeploymentResponse`?

File: internal/internal_workflow_execution_options.go
Comment (cretz) [reply internal/internal_workflow_execution_options.go:50]: We need to decide now how we're going to to do partial updates because `UpdateActivityOptions` (or `UpdateActivityOptionsById`) is coming with the same problem. I would do partial updates in the way the language handles it best. In Go unfortunately there is not one way, but this `UpdatedFields` is probably the least common and IMO we should not expose it.

I would either recommend using pointers or making an update options builder. Pointers works fine.

> I can see people reusing a data structure with the pointers, and making mistakes because they did not clear all of them

We should not use existing data structures in a partial update model because of this concern. So every data type for updating should not reuse the non-updating form on partial update things (this is not a full PUT-with-conflict-token type of API). So either inline the options update fields here or make the data type update-options specific.

File: .github/workflows/ci.yml
Comment (cretz) [inline .github/workflows/ci.yml:93]: We should not merge unless/until we can remove this flag IMO. If that means we need an RC release of CLI we can, though hopefully CLI for 1.26 is released before this so users of this SDK can test too.

Comment (antlai-temporal) [reply .github/workflows/ci.yml:93]: We will have an RC for CLI and run  integration test in CI before merging, see: 
https://github.com/temporalio/sdk-go/blob/7e976b2a877e1fd95c98750a8a0e955989e0fd4f/internal/cmd/build/main.go#L143-L145

This flag only disables the tests for the docker image, which unfortunately only gets built for real releases, but the integration tests do run in against that CLI with -dev-server:
https://github.com/temporalio/sdk-go/blob/7e976b2a877e1fd95c98750a8a0e955989e0fd4f/.github/workflows/ci.yml#L47-L59

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:222]: I think fairly confident, that API just returns info on a particular deployment, and we control the struct `DeploymentInfo`, since it is only used for `Describe` and `GetCurrent`, so we could extend it if needed...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:237]: Same argument as above, I wanted it to be symmetrical with `Describe`,  I think we can extend the struct `DeploymentInfo` without needing a wrapper...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:140]: It is an optimization to reduce the number of roundtrips vs memory available. Frankly, doubt many people will use it, but it was useful at least for testing (forcing several roundtrips...).

File: client/client.go
Comment (antlai-temporal) [reply client/client.go:954]: Good point, the same info is available in the response protos but it does not return that type...

File: client/client.go
Comment (antlai-temporal) [reply client/client.go:956]: Yes, I agree is a bit confusing, I'm just trying to be consistent with the API that would have been better with `*Properties`

Comment (antlai-temporal) [reply client/client.go:956]: Yes, I also think that "update" is too tainted with our signal/update/... and it is not always clear it is a verb coming from CRUD, and not a noun (or adjective) from our updates...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:40]: That value should always come from the server. Reachability is expensive and it is only run every so often, so it could be the server not having a cached value, and deciding not to run it either... I'll ask the server folks...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:71]: Thanks, fixed.

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:103]: It was `Current`, @ShahabT asked to change it to `IsCurrent`, it is a bool so it should be clear, but I don't have strong opinions...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:106]: Thanks, much better plural.

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:155]: Yes, that was my thinking, just "Info" was too overloaded...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:233]: Getting Reachability info is a very expensive operation, so it may be stale and/or rate limited, and that's why they provided a "light" version with `Describe` that only returns the easy bits... I will add a comment.

File: internal/internal_deployment_client.go
Comment (antlai-temporal) [reply internal/internal_deployment_client.go:61]: Done

File: internal/internal_deployment_client.go
Comment (antlai-temporal) [reply internal/internal_deployment_client.go:177]: Good point, thanks. I added it inside `paginate()`, `List` just returns the iterator struct...

File: internal/internal_workflow_execution_options.go
Comment (antlai-temporal) [reply internal/internal_workflow_execution_options.go:50]: I'm rewriting it with pointers. We only allow "top level object" partial updates, not arbitrary zipping, so I think that as long as the top level container is a different type we should be ok blocking reuse...

File: client/client.go
Comment (cretz) [reply client/client.go:961]: The problem is we have `UpdateActivityOptions` coming (or `UpdateActivityOptionsById`). But I still think what's here is probably the best.

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:103]: It is a bit strange to use `Is` prefixes in Go fields for these kinds of values IMO (unlike Java where of course you will)

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:140]: Hrmm, wonder if we should remove it. Will defer to others' opinions here.

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:235]: Hrmm, to me it sounds like overlapping concerns and to obtain reachability should just be an option on describe. I assume this was discussed at the API level and decided to separate `DescribeDeployment` and `GetDeploymentReachability` instead of just make this an option on `Describe` (like we have on `DescribeTaskQueue`)? If it was discussed, can you share the reasoning there?

File: internal/internal_deployment_client.go
Comment (cretz) [reply internal/internal_deployment_client.go]: Right, I think for schedules it was confirmed. If we can just get server team confirmation that no matter what is in `ListDeploymentsRequest` (e.g. `series_name`) it is impossible to get an empty page with a next page token we're good.

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:222]: For instance, like with `DescribeTaskQueue`, what if we decided to add the ability to request more things (which arguably should have been done with reachability information)? IMO in a perfect world `Describe` would return a `DeploymentDescription` and that description would have info and (opt-in) reachability info today and can grow to whatever else in the future.

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:239]: We often add things to the gRPC response message and not the single info field within. We have done this before on other calls and regret reusing types as the results on these methods because we are forced to make more.

File: .github/workflows/ci.yml
Comment (cretz) [reply .github/workflows/ci.yml:93]: Works for me. I wonder it the 1.26 release (and subsequent CLI release) will come before the RC.

File: internal/internal_workflow_client.go
Comment (cretz) [inline internal/internal_workflow_client.go:1072]: I think we should fail if there are no changes and/or a missing workflow ID (general validation like this applies to other calls too of course)

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:235]: Yes, it was discussed, they can only do rate limiting at the method level, and that's why they want a separate method.

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:222]: If we wanted we could make reachability an optional field in DeploymentInfo, I think extending is not a problem. Note that in versioning-1 there was a separate method for GetReachability, they combine it with DescribeTaskQueue in versioning-2 and that created many problems because that was a popular method, we wanted to abstract away from TQ,  and they could not rate limit cleanly, so we are back to the versioning-1 approach...
But I see your point that reachability is just info about the deployment, maybe we could group both with a flag in the cli?

Alternatively, I could wrap this in the standard "*Response" boring proto and kick the can down the road...

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:239]: OK, wrapping the responses

File: internal/internal_workflow_client.go
Comment (antlai-temporal) [reply internal/internal_workflow_client.go:1072]: Adding more validation

File: internal/deployment_client.go
Comment (antlai-temporal) [reply internal/deployment_client.go:140]: We have it in ScheduleListOptions, I think we should be consistent...

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:235]: Hrmm, that is unfortunate to burden a user API all the way out through the SDK with our server limitations, but ok. I am guessing they decided it was ok not to need to separately rate limit the different `DescribeTaskQueueRequest` options.

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:222]: I was thinking this would return a `DeploymentDescription` that would have `Info` as its only field today. That way it can grow if the gRPC API does. If we wanted to though, we could combine this call with `GetReachability` and then `DeploymentDescription` would have two fields, but I understand from the other thread it was decided elsewhere to keep them separate unfortunately because of some internal server limitation.

Comment (antlai-temporal) [reply internal/deployment_client.go:222]: OK, adding a DeploymentDescription and also wrapping the "*Response"

File: internal/deployment_client.go
Comment (cretz) [reply internal/deployment_client.go:140]: We don't for `GetWorkflowHistory` which returns an iterator (even though we could), so we're going to be inconsistent regardless. But I am ok with leaving it.

File: internal/internal_deployment_client.go
Comment (carlydf) [inline internal/internal_deployment_client.go:62]: `ListDeployments` uses `visibility.ListWorkflowExecutions` under the hood directly, the same way that `ListSchedules` does, so the contents of a page and the `nextPageToken` are handled the same for `ListDeployments` and `ListSchedules`. Because of this, I believe that there is no risk of an empty page followed by a non-empty page.

Comment (carlydf) [reply internal/internal_deployment_client.go:62]: In the elastic search visibility store implementation, next page token is only non-nil if the number of entries on the page == pageSize: https://github.com/temporalio/temporal/blob/main/common/persistence/visibility/store/elasticsearch/visibility_store.go#L839-L848

Comment (carlydf) [reply internal/internal_deployment_client.go:62]: @cretz @antlai-temporal hope this helps

Comment (carlydf) [reply internal/internal_deployment_client.go:62]: same in the sql visibility store implementation, next page token is only non-nil if the number of entries on the page == pageSize: https://github.com/temporalio/temporal/blob/main/common/persistence/visibility/store/sql/visibility_store.go#L233-L248

Comment (cretz) [reply internal/internal_deployment_client.go:62]: > I believe that there is no risk of an empty page followed by a non-empty page

Sounds good. Just wanted to make sure there was no post-visibility item filtering done server side that would affect this.

**Review by cretz — COMMENTED** [review]
Looks good, just waiting for CLI to be updated properly

**Comment by antlai-temporal** [comment]
@cretz Integration tests against the new cli ok...

**Review by cretz — APPROVED** [review]

**Pull request merged**
